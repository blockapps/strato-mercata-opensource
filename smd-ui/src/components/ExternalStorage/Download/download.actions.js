export const OPEN_DOWNLOAD_MODAL = 'OPEN_DOWNLOAD_MODAL';
export const CLOSE_DOWNLOAD_MODAL = 'CLOSE_DOWNLOAD_MODAL';
export const DOWNLOAD_REQUEST = 'DOWNLOAD_REQUEST';
export const DOWNLOAD_SUCCESS = 'DOWNLOAD_SUCCESS';
export const DOWNLOAD_FAILURE = 'DOWNLOAD_FAILURE';
export const RESET_ERROR = 'RESET_ERROR';
export const CLEAR_URL = 'CLEAR_URL';

export const openDownloadModal = function () {
  return {
    type: OPEN_DOWNLOAD_MODAL
  }
}

export const closeDownloadModal = function () {
  return {
    type: CLOSE_DOWNLOAD_MODAL
  }
}

export const downloadRequest = function (contractAddress) {
  return {
    type: DOWNLOAD_REQUEST,
    contractAddress
  }
}

export const downloadSuccess = function (url) {
  return {
    type: DOWNLOAD_SUCCESS,
    url: url
  }
}

export const downloadFailure = function (error) {
  return {
    type: DOWNLOAD_FAILURE,
    error
  }
}

export const resetError = function () {
  return {
    type: RESET_ERROR
  }
}

export const clearUrl = function () {
  return {
    type: CLEAR_URL
  }
}