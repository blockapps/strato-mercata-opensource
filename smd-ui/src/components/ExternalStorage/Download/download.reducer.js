import { OPEN_DOWNLOAD_MODAL, CLOSE_DOWNLOAD_MODAL, DOWNLOAD_FAILURE, RESET_ERROR, DOWNLOAD_SUCCESS, CLEAR_URL } from "./download.actions";

const initialState = {
  isOpen: false,
  error: null,
  url: null
}

const reducer = function (state = initialState, action) {
  switch (action.type) {
    case OPEN_DOWNLOAD_MODAL:
      return {
        ...state,
        isOpen: true
      }
    case CLOSE_DOWNLOAD_MODAL:
      return {
        ...state,
        isOpen: false
      }
    case DOWNLOAD_SUCCESS:
      return {
        ...state,
        url: action.url
      }
    case DOWNLOAD_FAILURE:
      return {
        ...state,
        error: action.error
      }
    case RESET_ERROR:
      return {
        ...state,
        error: null
      }
    case CLEAR_URL:
      return {
        ...state,
        url: null,
        error: null
      }
    default:
      return state;
  }
}

export default reducer;