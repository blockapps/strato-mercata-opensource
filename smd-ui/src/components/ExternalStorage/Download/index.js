import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Dialog, Button, Intent } from '@blueprintjs/core';
import { Field, reduxForm } from 'redux-form';
import mixpanelWrapper from '../../../lib/mixpanelWrapper';
import validate from './validate';
import { closeDownloadModal, downloadRequest, resetError, clearUrl } from './download.actions';
import { toasts } from '../../Toasts';

class Download extends Component {

  constructor(props) {
    super(props);
    this.state = { errors: null }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.downloadError) {
      toasts.show({ message: nextProps.downloadError });
      this.props.resetError();
    }
    if (nextProps.downloadUrl) {
      window.open(nextProps.downloadUrl, "_parent")
      this.props.clearUrl();
    }
  }

  submit = (values) => {
    let errors = validate(values);
    this.setState({ errors });

    if (!Object.values(errors).length) {
      this.props.downloadRequest(values.contractAddress);
    }
  }

  errorMessageFor(fieldName) {
    if (this.state.errors && this.state.errors[fieldName]) {
      return this.state.errors[fieldName];
    }
    return null;
  }

  renderDownloadForm() {
    return (
      <form>
        <div className="pt-dialog-body upload-form">

          <div className="row">
            <div className="col-sm-4 text-right">
              <label className="pt-label smd-pad-4">
                Contract Address
            </label>
            </div>
            <div className="col-sm-8 smd-pad-4">
              <Field
                name="contractAddress"
                component="input"
                type="text"
                placeholder="Contract Address"
                className="pt-input form-width"
                tabIndex="1"
                required
              />
              <br /><span className="error-text">{this.errorMessageFor('contractAddress')}</span>
            </div>
          </div>

          <div className="pt-dialog-footer">
            <div className="pt-dialog-footer-actions button-center smd-margin-8">
              <Button
                intent={Intent.PRIMARY}
                onClick={this.props.handleSubmit(this.submit)}
                text="Download"
              />
            </div>
          </div>
        </div>
      </form>
    );
  }

  render() {
    return (
      <div>
        <Dialog
          isOpen={this.props.isOpen}
          onClose={() => {
            mixpanelWrapper.track('close_download_modal');
            this.props.closeDownloadModal();
            this.props.reset();
          }}
          iconName='pt-icon-download'
          title='Download'
          className="pt-dark"
        >
          {this.renderDownloadForm()}
        </Dialog>
      </div>
    )
  }
}

export function mapStateToProps(state) {
  return {
    isOpen: state.download.isOpen,
    downloadError: state.download.error,
    downloadUrl: state.download.url
  };
}

const formed = reduxForm({ form: 'verify-form' })(Download);
const connected = connect(
  mapStateToProps,
  {
    closeDownloadModal,
    downloadRequest,
    resetError,
    clearUrl
  }
)(formed);

export default withRouter(connected);