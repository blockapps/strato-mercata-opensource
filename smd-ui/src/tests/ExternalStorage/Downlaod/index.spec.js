import React from 'react';
import Download, { mapStateToProps } from '../../../components/ExternalStorage/Download';
import { createStore, combineReducers } from 'redux';
import { reducer as formReducer } from 'redux-form';

describe('Download: index', () => {
  let store

  beforeEach(() => {
    store = createStore(combineReducers({ form: formReducer }));
  });

  describe('render component', () => {

    test('without values', () => {
      const props = {
        isOpen: false,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      expect(wrapper.debug()).toMatchSnapshot();
    });

    test('with values', () => {
      const props = {
        isOpen: true,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      expect(wrapper.debug()).toMatchSnapshot();
    });

  });

  describe('Modal', () => {

    test('onClose', () => {
      const props = {
        isOpen: false,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.find('Dialog').get(0).props.onClose();
      expect(props.closeDownloadModal).toHaveBeenCalled();
      expect(props.closeDownloadModal).toHaveBeenCalledTimes(1);
    });

    test('Form', () => {
      const props = {
        isOpen: false,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      const field = wrapper.find('Field');
      field.at(0).simulate('change', { target: { value: '23fe0e0ab7bd95fbcdff877660c595d24c6dcf5c' } });
      wrapper.find('Button').first().simulate('click');
      expect(props.handleSubmit).toHaveBeenCalled();
    });

    test('Submit with empty value', () => {
      const props = {
        isOpen: true,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.instance().submit({});
      expect(wrapper.state()).toMatchSnapshot();
    });

    test('Submit with values', () => {
      const props = {
        isOpen: true,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.instance().submit({ contractAddress: '23fe0e0ab7bd95fbcdff877660c595d24c6dcf5c' });
      expect(props.downloadRequest).toHaveBeenCalled();
      expect(props.downloadRequest).toHaveBeenCalledTimes(1);
    });
  });

  describe('componentWillReceiveProps', () => {

    test('with values', () => {
      const props = {
        isOpen: true,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      // true case
      wrapper.instance().componentWillReceiveProps({ downloadError: 'error', downloadUrl: 'https://strato-external-storage.s3.amazonaws.com/1529915329415-widescreen.jpeg' });
      expect(props.resetError).toHaveBeenCalled();
      expect(props.resetError).toHaveBeenCalledTimes(1);
      expect(props.clearUrl).toHaveBeenCalled();
      expect(props.clearUrl).toHaveBeenCalledTimes(1);
    });

    test('without values', () => {
      const props = {
        isOpen: true,
        downloadError: null,
        downloadUrl: null,
        clearUrl: jest.fn(),
        downloadRequest: jest.fn(),
        handleSubmit: jest.fn(),
        closeDownloadModal: jest.fn(),
        reset: jest.fn(),
        resetError: jest.fn(),
        store: store
      }

      const wrapper = shallow(
        <Download.WrappedComponent {...props} />
      ).dive().dive().dive();

      wrapper.instance().componentWillReceiveProps({});
      expect(props.resetError).not.toHaveBeenCalled();
      expect(props.clearUrl).not.toHaveBeenCalled();
    });

  });

  describe('mapStateToProps', () => {

    test('without values', () => {
      const state = {
        download: {
          isOpen: false,
          error: null,
          url: null
        }
      }

      expect(mapStateToProps(state)).toMatchSnapshot();
    });

    test('with values', () => {
      const state = {
        download: {
          isOpen: true,
          error: 'error',
          url: 'https://strato-external-storage.s3.amazonaws.com/1529905060401-widescreen.jpeg'
        }
      }

      expect(mapStateToProps(state)).toMatchSnapshot();
    });

  });

});