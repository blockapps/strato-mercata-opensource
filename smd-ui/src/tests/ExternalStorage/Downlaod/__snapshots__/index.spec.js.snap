// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Download: index Modal Submit with empty value 1`] = `
Object {
  "errors": Object {
    "contractAddress": "Can't be blank",
  },
}
`;

exports[`Download: index mapStateToProps with values 1`] = `
Object {
  "downloadError": "error",
  "downloadUrl": "https://strato-external-storage.s3.amazonaws.com/1529905060401-widescreen.jpeg",
  "isOpen": true,
}
`;

exports[`Download: index mapStateToProps without values 1`] = `
Object {
  "downloadError": null,
  "downloadUrl": null,
  "isOpen": false,
}
`;

exports[`Download: index render component with values 1`] = `
"<div>
  <Blueprint.Dialog isOpen={true} onClose={[Function: onClose]} iconName=\\"pt-icon-download\\" title=\\"Download\\" className=\\"pt-dark\\" canOutsideClickClose={true}>
    <form>
      <div className=\\"pt-dialog-body upload-form\\">
        <div className=\\"row\\">
          <div className=\\"col-sm-4 text-right\\">
            <label className=\\"pt-label smd-pad-4\\">
              Contract Address
            </label>
          </div>
          <div className=\\"col-sm-8 smd-pad-4\\">
            <Field name=\\"contractAddress\\" component=\\"input\\" type=\\"text\\" placeholder=\\"Contract Address\\" className=\\"pt-input form-width\\" tabIndex=\\"1\\" required={true} />
            <br />
            <span className=\\"error-text\\" />
          </div>
        </div>
        <div className=\\"pt-dialog-footer\\">
          <div className=\\"pt-dialog-footer-actions button-center smd-margin-8\\">
            <Blueprint.Button intent={0} onClick={[undefined]} text=\\"Download\\" />
          </div>
        </div>
      </div>
    </form>
  </Blueprint.Dialog>
</div>"
`;

exports[`Download: index render component without values 1`] = `
"<div>
  <Blueprint.Dialog isOpen={false} onClose={[Function: onClose]} iconName=\\"pt-icon-download\\" title=\\"Download\\" className=\\"pt-dark\\" canOutsideClickClose={true}>
    <form>
      <div className=\\"pt-dialog-body upload-form\\">
        <div className=\\"row\\">
          <div className=\\"col-sm-4 text-right\\">
            <label className=\\"pt-label smd-pad-4\\">
              Contract Address
            </label>
          </div>
          <div className=\\"col-sm-8 smd-pad-4\\">
            <Field name=\\"contractAddress\\" component=\\"input\\" type=\\"text\\" placeholder=\\"Contract Address\\" className=\\"pt-input form-width\\" tabIndex=\\"1\\" required={true} />
            <br />
            <span className=\\"error-text\\" />
          </div>
        </div>
        <div className=\\"pt-dialog-footer\\">
          <div className=\\"pt-dialog-footer-actions button-center smd-margin-8\\">
            <Blueprint.Button intent={0} onClick={[undefined]} text=\\"Download\\" />
          </div>
        </div>
      </div>
    </form>
  </Blueprint.Dialog>
</div>"
`;
